contrib/colab/mnist-training-xrt-1-15.ipynb

```diff
@@ -0,0 +1,312 @@
+        "  print('[{}] Accuracy={:.2f}%\\n'.format(device, accuracy))\n",
+        "  return accuracy, data, pred, target"
+      ],
+      "execution_count": 3,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "Devices: ['xla:1', 'xla:2', 'xla:3', 'xla:4', 'xla:5', 'xla:6', 'xla:7', 'xla:8']\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "EBRzfYprFDbo",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "# Start training\n",
+        "data, target, pred = None, None, None\n",
+        "for epoch in range(1, num_epochs + 1):\n",
+        "  model_parallel(train_loop_fn, train_loader)\n",
+        "  results = model_parallel(test_loop_fn, test_loader)\n",
+        "  results = np.array(results)  # for ease of slicing\n",
+        "  accuracies, data, pred, target = \\\n",
+        "    [results[:,i] for i in range(results.shape[-1])]  \n",
+        "  accuracy = sum(accuracies) / len(accuracies)\n",
+        "  print('[Epoch {}] Global accuracy: {:.2f}%'.format(epoch, accuracy))\n",
+        "  if metrics_debug:\n",
+        "    print(torch_xla._XLAC._xla_metrics_report())"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "zaaW5RWAJQO2",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Visualize Predictions"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "5bHyp9HWJPoz",
+        "colab_type": "code",
+        "outputId": "c727db4a-e2b0-4606-c0d1-1b384d3050ea",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 593
+        }
+      },
+      "source": [
+        "%matplotlib inline\n",
+        "from matplotlib import pyplot as plt\n",
+        "\n",
+        "# Retrieve tensors that is on TPU cores\n",
+        "M = 4\n",
+        "N = 6\n",
+        "images, labels, preds = data[0][:N*M].cpu(), \\\n",
+        "  target[0][:N*M].cpu(), pred[0][:N*M].cpu()\n",
+        "\n",
+        "fig, axes = plt.subplots(M, N, figsize=(11, 9))\n",
+        "fig.suptitle('Correct / Predicted Labels (Red text for incorrect ones)')\n",
+        "\n",
+        "for i, ax in enumerate(fig.axes):\n",
+        "  img, label, prediction = images[i], labels[i], preds[i]\n",
+        "  img = inv_norm(img)\n",
+        "  img = img.squeeze() # [1,Y,X] -> [Y,X]\n",
+        "  ax.axis('off')\n",
+        "  label, prediction = label.item(), prediction.item()\n",
+        "  if label == prediction:\n",
+        "    ax.set_title(u'\\u2713', color='blue', fontsize=22)\n",
+        "  else:\n",
+        "    ax.set_title(\n",
+        "        'X {}/{}'.format(label, prediction), color='red')\n",
+        "  ax.imshow(img)"
+      ],
+      "execution_count": 5,
+      "outputs": [
```

Comment:
I would not capture the outputs. Bloat the Colab.
Same all over these Colabs.

Reply:
Done.